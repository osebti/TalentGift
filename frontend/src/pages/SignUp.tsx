import { Link, useNavigate } from "react-router-dom";
import { useRef } from "react";
import config from "../../../config.js";

// components
import Button from "@/components/Button";
import TextBox from "@/components/TextBox";
import NavBar from "@/components/NavBar";
import { setGlobalState, useGlobalState } from "@/state/global";

interface TextboxProps {
  label: string;
  placeholder: string;
  type: string;
  ref: { current: HTMLInputElement | null };
}

const SignSwapSection = () => {
  const link = "/sign-up";
  const text = "Don't have an account?";
  const button = "Sign Up";
  return (
    <span
      id="sign signup container"
      className="flex align-middle justify-center"
    >
      <Link to={link} id="signSwap text" className="text-lg justify-self-end">
        {text}
      </Link>
      <Link to={link} id="signSwap button" className="text-lg font-bold pl-2">
        {button}
      </Link>
    </span>
  );
};

export default function SignUpPage() {
  const emailRef = useRef<HTMLInputElement>(null);
  const passwordRef = useRef<HTMLInputElement>(null);
  const rememberRef = useRef<HTMLInputElement>(null);
  const navigate = useNavigate();

  const [user] = useGlobalState("user");
  const [org] = useGlobalState("org");

  const textboxes: TextboxProps[] = [
    {
      label: "Email",
      placeholder: "Enter your email",
      type: "email",
      ref: emailRef,
    },
    {
      label: "Retype Email",
      placeholder: "Enter your email",
      type: "email",
      ref: emailRef,
    },
    {
      label: "Password",
      placeholder: "Enter your password",
      type: "password",
      ref: passwordRef,
    },
  ];

  const loginOnClick = () => {
    console.log("Login button clicked");
    if (emailRef.current && passwordRef.current) {
      console.log(
        `Received Username: ${emailRef.current.value} and Password: ${passwordRef.current.value}`
      );
      if (rememberRef.current) {
        console.log(`Remember me: ${rememberRef.current.checked}`);
      }

      // Send login request to backend
      const email = emailRef.current.value;
      const password = passwordRef.current.value;
      fetch(`http://${config.hostname}/signUp`, {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ email, password }),
      })
        .then((res) => {
          if (!res.ok) {
            throw Error("Internal Error");
          } else {
            // Signed up successfully
            window.alert("Account successfully created!");
            return res.json();
          }
        })
        .then((data) => {
          // Update global state and localStorage (oid is 0 since user has not joined organization on sign up)
          setGlobalState("user", {
            ...user,
            manager: false,
            uid: String(data.user.uid),
            firstName: data.user.firstname,
            lastName: data.user.lastname,
          });
          setGlobalState("org", { ...org, oid: "0" });
          localStorage.setItem("uid", String(data.user.uid));
          localStorage.setItem("firstName", data.user.firstname);
          localStorage.setItem("lastName", data.user.lastname);
          localStorage.setItem("email", data.user.email);
          localStorage.setItem("manager", "false");
          localStorage.setItem("oid", "0");
          localStorage.setItem("notifications", "false");
          navigate("/dashboard");
        });
    }
  };

  return (
    <div className="flex flex-col gap-y-16 h-screen">
      <NavBar hideButton={true} className="mx-4 mt-4 md:mx-8" />
      <div className="flex-grow grid lg:grid-cols-2 place-items-center gap-x-32 px-4 md:px-8">
        <div
          id="sign container"
          className="flex flex-row flex-wrap justify-center items-center col-start-1"
        >
          <div
            id="sign content container"
            className="flex flex-col justify-center items-center gap-4 md:gap-8"
          >
            <p
              id="sign content header"
              className="text-4xl md:text-6xl font-bold text-center whitespace-nowrap"
            >
              Getting Started
            </p>
            <p id="sign content text" className="text-lg text-center">
              Unite to foster an environment where creativity, collaboration,
              and mutual respect flourish
            </p>
            {textboxes.map((textbox, index) => (
              <TextBox
                id={textbox.label}
                key={index}
                textRef={textbox.ref}
                label={textbox.label}
                placeholder={textbox.placeholder}
                type={textbox.type}
              />
            ))}
            <Button
              id="Sign up"
              label="Sign up"
              style="btn-secondary w-full mt-4 md:mt-0"
              onClick={loginOnClick}
              navigate={false}
            />
            <SignSwapSection />
          </div>
        </div>
        <div className="h-full col-start-2 hidden lg:flex">
          <img
            alt="Sign Up Illustration"
            src="sign_in_up.svg"
            className="object-contain"
          />
        </div>
      </div>
    </div>
  );
}
