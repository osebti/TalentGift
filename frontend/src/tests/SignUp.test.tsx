//#region imports
import { cleanup, fireEvent, render, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import fetchMock from "jest-fetch-mock";

import SignUp from "@/pages/SignUp";
import { BrowserRouter } from "react-router-dom";
//#endregion

//#region test setup
fetchMock.enableMocks();
global.alert = jest.fn();
console.error = jest.fn();
console.warn = jest.fn();
console.log = jest.fn();

beforeEach(() => {
  fetchMock.resetMocks();
});

afterEach(cleanup);
//#endregion

//#region functions and objects used for testing
function renderSignUpPage() {
  return render(
    <BrowserRouter>
      <SignUp />
    </BrowserRouter>
  );
}
//#endregion

//#region tests
describe("Rendering the SignUp Page", () => {
  it("renders correctly", () => {
    const page = renderSignUpPage();
    expect(page).toMatchSnapshot();
  });

  it("renders the email input field", () => {
    const { queryByTestId } = renderSignUpPage();
    expect(queryByTestId("Email textbox")).not.toBeNull();
  });

  it("renders the retype email input field", () => {
    const { queryByTestId } = renderSignUpPage();
    expect(queryByTestId("Retype Email textbox")).not.toBeNull();
  });

  it("renders the password input field", () => {
    const { queryByTestId } = renderSignUpPage();
    expect(queryByTestId("Password textbox")).not.toBeNull();
  });

  it("renders the sign up button", () => {
    const { queryByTestId } = renderSignUpPage();
    expect(queryByTestId("Sign up-button")).not.toBeNull();
  });
});

describe("Signing up", () => {
  it("sign up with valid email and password", async () => {
    const { queryByTestId } = renderSignUpPage();
    // mock successful sign up request
    fetchMock.mockResponseOnce(
      JSON.stringify({
        status: 0,
        user: {
          uid: 1,
        },
        message: "User sucessfully added!",
      }),
      { status: 200, headers: { "Content-Type": "application/json" } }
    );

    // enter an email and password and click the sign up button
    const emailInput = queryByTestId("Email textbox");
    const passwordInput = queryByTestId("Password textbox");
    const signUpButton = queryByTestId("Sign up-button");

    if (emailInput != null && passwordInput != null && signUpButton != null) {
      fireEvent.change(emailInput, { target: { value: "test@gmail.com" } });
      fireEvent.change(passwordInput, { target: { value: "12345test" } });
      fireEvent.click(signUpButton);
    }

    // check if a fetch request was sent and the user was properly redirected
    await waitFor(() => {
      expect(fetchMock).toHaveBeenCalledTimes(1);
      expect(global.alert).toBeCalled();
      expect(global.window.location.href).toContain("/dashboard");
    });
  });
});
//#endregion
