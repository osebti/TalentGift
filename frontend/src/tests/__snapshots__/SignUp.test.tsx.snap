// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering the SignUp Page renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col gap-y-16 h-screen"
      >
        <div
          class="mx-4 mt-4 md:mx-8 flex justify-between items-center"
        >
          <a
            class="font-[1000] text-darkPurple text-3xl bg-transparent border-0"
            href="/"
          >
            TG
          </a>
          <div
            class="gap-x-4 hidden"
          >
            <a
              class="btn btn-md rounded-full capitalize font-normal text-white text-sm md:text-base btn-accent btn-sm md:btn-md px-4 md:px-8"
              data-testid="Sign In-button"
              href="/sign-in"
              id="Sign In-button"
            >
              Sign In
            </a>
            <a
              class="btn btn-md rounded-full capitalize font-normal text-white text-sm md:text-base btn-primary btn-sm md:btn-md px-4 md:px-8"
              data-testid="Sign Up-button"
              href="/sign-up"
              id="Sign Up-button"
            >
              Sign Up
            </a>
          </div>
        </div>
        <div
          class="flex-grow grid lg:grid-cols-2 place-items-center gap-x-32 px-4 md:px-8"
        >
          <div
            class="flex flex-row flex-wrap justify-center items-center col-start-1"
            id="sign container"
          >
            <div
              class="flex flex-col justify-center items-center gap-4 md:gap-8"
              id="sign content container"
            >
              <p
                class="text-4xl md:text-6xl font-bold text-center whitespace-nowrap"
                id="sign content header"
              >
                Getting Started
              </p>
              <p
                class="text-lg text-center"
                id="sign content text"
              >
                Unite to foster an environment where creativity, collaboration, and mutual respect flourish
              </p>
              <div
                class="flex-col justify-start items-start w-full"
                id="Email-container"
              >
                <div
                  class="text-start text-black text-xl font-normal pb-2"
                  data-testid="Email header"
                  id="Email-header"
                >
                  Email
                </div>
                <input
                  class="text-black input outline outline-gray-300 text-xl font-normal bg-transparent w-full px-3 py-2.5 rounded-md justify-start items-center inline-flex focus:outline-darkPurple"
                  data-testid="Email textbox"
                  id="Email-textbox"
                  placeholder="Enter your email"
                  type="email"
                />
              </div>
              <div
                class="flex-col justify-start items-start w-full"
                id="Retype Email-container"
              >
                <div
                  class="text-start text-black text-xl font-normal pb-2"
                  data-testid="Retype Email header"
                  id="Retype Email-header"
                >
                  Retype Email
                </div>
                <input
                  class="text-black input outline outline-gray-300 text-xl font-normal bg-transparent w-full px-3 py-2.5 rounded-md justify-start items-center inline-flex focus:outline-darkPurple"
                  data-testid="Retype Email textbox"
                  id="Retype Email-textbox"
                  placeholder="Enter your email"
                  type="email"
                />
              </div>
              <div
                class="flex-col justify-start items-start w-full"
                id="Password-container"
              >
                <div
                  class="text-start text-black text-xl font-normal pb-2"
                  data-testid="Password header"
                  id="Password-header"
                >
                  Password
                </div>
                <input
                  class="text-black input outline outline-gray-300 text-xl font-normal bg-transparent w-full px-3 py-2.5 rounded-md justify-start items-center inline-flex focus:outline-darkPurple"
                  data-testid="Password textbox"
                  id="Password-textbox"
                  placeholder="Enter your password"
                  type="password"
                />
              </div>
              <button
                class="btn btn-md rounded-full capitalize font-normal text-white text-sm md:text-base btn-secondary w-full mt-4 md:mt-0"
                data-testid="Sign up-button"
                id="Sign up-button"
              >
                Sign up
              </button>
              <span
                class="flex align-middle justify-center"
                id="sign signup container"
              >
                <a
                  class="text-lg justify-self-end"
                  href="/sign-up"
                  id="signSwap text"
                >
                  Don't have an account?
                </a>
                <a
                  class="text-lg font-bold pl-2"
                  href="/sign-up"
                  id="signSwap button"
                >
                  Sign Up
                </a>
              </span>
            </div>
          </div>
          <div
            class="h-full col-start-2 hidden lg:flex"
          >
            <img
              alt="Sign Up Illustration"
              class="object-contain"
              src="sign_in_up.svg"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col gap-y-16 h-screen"
    >
      <div
        class="mx-4 mt-4 md:mx-8 flex justify-between items-center"
      >
        <a
          class="font-[1000] text-darkPurple text-3xl bg-transparent border-0"
          href="/"
        >
          TG
        </a>
        <div
          class="gap-x-4 hidden"
        >
          <a
            class="btn btn-md rounded-full capitalize font-normal text-white text-sm md:text-base btn-accent btn-sm md:btn-md px-4 md:px-8"
            data-testid="Sign In-button"
            href="/sign-in"
            id="Sign In-button"
          >
            Sign In
          </a>
          <a
            class="btn btn-md rounded-full capitalize font-normal text-white text-sm md:text-base btn-primary btn-sm md:btn-md px-4 md:px-8"
            data-testid="Sign Up-button"
            href="/sign-up"
            id="Sign Up-button"
          >
            Sign Up
          </a>
        </div>
      </div>
      <div
        class="flex-grow grid lg:grid-cols-2 place-items-center gap-x-32 px-4 md:px-8"
      >
        <div
          class="flex flex-row flex-wrap justify-center items-center col-start-1"
          id="sign container"
        >
          <div
            class="flex flex-col justify-center items-center gap-4 md:gap-8"
            id="sign content container"
          >
            <p
              class="text-4xl md:text-6xl font-bold text-center whitespace-nowrap"
              id="sign content header"
            >
              Getting Started
            </p>
            <p
              class="text-lg text-center"
              id="sign content text"
            >
              Unite to foster an environment where creativity, collaboration, and mutual respect flourish
            </p>
            <div
              class="flex-col justify-start items-start w-full"
              id="Email-container"
            >
              <div
                class="text-start text-black text-xl font-normal pb-2"
                data-testid="Email header"
                id="Email-header"
              >
                Email
              </div>
              <input
                class="text-black input outline outline-gray-300 text-xl font-normal bg-transparent w-full px-3 py-2.5 rounded-md justify-start items-center inline-flex focus:outline-darkPurple"
                data-testid="Email textbox"
                id="Email-textbox"
                placeholder="Enter your email"
                type="email"
              />
            </div>
            <div
              class="flex-col justify-start items-start w-full"
              id="Retype Email-container"
            >
              <div
                class="text-start text-black text-xl font-normal pb-2"
                data-testid="Retype Email header"
                id="Retype Email-header"
              >
                Retype Email
              </div>
              <input
                class="text-black input outline outline-gray-300 text-xl font-normal bg-transparent w-full px-3 py-2.5 rounded-md justify-start items-center inline-flex focus:outline-darkPurple"
                data-testid="Retype Email textbox"
                id="Retype Email-textbox"
                placeholder="Enter your email"
                type="email"
              />
            </div>
            <div
              class="flex-col justify-start items-start w-full"
              id="Password-container"
            >
              <div
                class="text-start text-black text-xl font-normal pb-2"
                data-testid="Password header"
                id="Password-header"
              >
                Password
              </div>
              <input
                class="text-black input outline outline-gray-300 text-xl font-normal bg-transparent w-full px-3 py-2.5 rounded-md justify-start items-center inline-flex focus:outline-darkPurple"
                data-testid="Password textbox"
                id="Password-textbox"
                placeholder="Enter your password"
                type="password"
              />
            </div>
            <button
              class="btn btn-md rounded-full capitalize font-normal text-white text-sm md:text-base btn-secondary w-full mt-4 md:mt-0"
              data-testid="Sign up-button"
              id="Sign up-button"
            >
              Sign up
            </button>
            <span
              class="flex align-middle justify-center"
              id="sign signup container"
            >
              <a
                class="text-lg justify-self-end"
                href="/sign-up"
                id="signSwap text"
              >
                Don't have an account?
              </a>
              <a
                class="text-lg font-bold pl-2"
                href="/sign-up"
                id="signSwap button"
              >
                Sign Up
              </a>
            </span>
          </div>
        </div>
        <div
          class="h-full col-start-2 hidden lg:flex"
        >
          <img
            alt="Sign Up Illustration"
            class="object-contain"
            src="sign_in_up.svg"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
